contracts/src/KyberElasticMultiQuoter.sol

```diff
@@ -0,0 +1,316 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+
+  Copyright 2022 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity 0.8.9;
+pragma experimental ABIEncoderV2;
+
+import "./interfaces/IKyberSwapElastic.sol";
+import "@kyberelastic/interfaces/IFactory.sol";
+import "@kyberelastic/libraries/TickMath.sol";
+import "@kyberelastic/libraries/SwapMath.sol";
+import "@kyberelastic/libraries/Linkedlist.sol";
+import "@kyberelastic/libraries/LiqDeltaMath.sol";
+import "@kyberelastic/periphery/libraries/PathHelper.sol";
+import "@uniswap/v3-core/contracts/libraries/LowGasSafeMath.sol";
+
+/// @title Provides quotes for multiple swap amounts
+/// @notice Allows getting the expected amount out or amount in for multiple given swap amounts without executing the swap
+contract KyberElasticMultiQuoter {
+    using SafeCast for uint256;
+    using LowGasSafeMath for int256;
+    using SafeCast for int128;
+
+    // temporary swap variables, some of which will be used to update the pool state
+    struct SwapData {
+        int256 specifiedAmount; // the specified amount (could be tokenIn or tokenOut)
+        int256 returnedAmount; // the opposite amout of sourceQty
+        uint160 sqrtP; // current sqrt(price), multiplied by 2^96
+        int24 currentTick; // the tick associated with the current price
+        int24 nextTick; // the next initialized tick
+        uint160 nextSqrtP; // the price of nextTick
+        bool isToken0; // true if specifiedAmount is in token0, false if in token1
+        bool isExactInput; // true = input qty, false = output qty
+        uint128 baseL; // the cached base pool liquidity without reinvestment liquidity
+        uint128 reinvestL; // the cached reinvestment liquidity
+        uint256 amountsIndex;
+    }
+
+    // the result of multiswap
+    struct MultiSwapResult {
+        // the gas estimate for each of swap amounts
+        uint256[] gasEstimates;
+        // the token0 delta for each swap amount, positive indicates sent and negative indicates receipt
+        int256[] amounts0;
+        // the token1 delta for each swap amount, positive indicates sent and negative indicates receipt
+        int256[] amounts1;
+    }
```

Comment:
Should we move this struct to a common file since its used in `UniswapV3MultiQuoter.sol` as well?